src/providers/virtualraster/qgsvirtualrasterprovider.cpp

```diff
@@ -0,0 +1,356 @@
+/***************************************************************************
+   qgsvirtualrasterprovider.cpp - QgsRasterDataProvider
+     --------------------------------------
+    Date                 : June 10, 2021
+    Copyright            : (C) 2021 by Francesco Bursi
+    email                : francesco.bursi
+***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#include "qgsvirtualrasterprovider.h"
+#include "qgsrastermatrix.h"
+#include "qgsrasterlayer.h"
+#include "qgsrasterprojector.h"
+
+//#define PROVIDER_KEY QStringLiteral( "virtualrasterprovider" )
+#define PROVIDER_KEY QStringLiteral( "virtualraster" )
+#define PROVIDER_DESCRIPTION QStringLiteral( "Virtual Raster data provider" )
+
+QgsVirtualRasterProvider::QgsVirtualRasterProvider( const QString &uri, const QgsDataProvider::ProviderOptions &providerOptions )
+  : QgsRasterDataProvider( uri, providerOptions )
+{
+  bool  ok;
+  QgsRasterDataProvider::VirtualRasterParameters decodedUriParams = QgsRasterDataProvider::decodeVirtualRasterProviderUri( uri, & ok );
+
+  if ( ok == false )
+  {
+    mValid = false;
+    return;
+  }
+
+  if ( ! decodedUriParams.crs.isValid() )
+  {
+    QgsDebugMsg( "crs is not valid" );
+    mValid = false;
+    return;
+  }
+  mCrs = decodedUriParams.crs;
+
+  if ( decodedUriParams.extent.isNull() )
+  {
+    QgsDebugMsg( "extent is null" );
+    mValid = false;
+    return;
+  }
+  mExtent = decodedUriParams.extent;
+
+  mWidth = decodedUriParams.width;
+  mHeight = decodedUriParams.height;
+
+  mFormulaString = decodedUriParams.formula;
+
+  mLastError.clear();
+  mCalcNode.reset( QgsRasterCalcNode::parseRasterCalcString( mFormulaString, mLastError ) );
+
+  if ( !mCalcNode )
+  {
+    mValid = false;
+    return;
+  }
+
+  QStringList rLayerDict = QgsRasterCalcNode::referencedLayerNames( mFormulaString );
+
+  QStringList rasterRefs;
+  const QList<const QgsRasterCalcNode *> rasterRefNodes = mCalcNode->findNodes( QgsRasterCalcNode::Type::tRasterRef );
+  for ( const QgsRasterCalcNode *r : rasterRefNodes )
+  {
+    QString layerRef( r->toString().remove( 0, 1 ) );
+
+    layerRef.chop( 1 );
```

Comment:
As mentioned earlier, also here - I think the raster references can be unquoted. Maybe it would be good to add another function to QgsRasterCalcNode to return "clean" version of raster reference without quotes so that the code is not spread in multiple places

Reply:
Ok, thank fo the advice, in fact there are some part that can be improved. I added a methods that should return a clean version of the the raster references and I used it in the above method and in the constructor of my provider, thank you